### ISSUE TITLE
Running nnUNet v2 in deterministic mode

### ISSUE THREAD

**Issue by joaco18** [MIC-DKFZ/nnUNet#1423]
Hi, thanks for this amazing paper/model/pipeline/library, I've been working with it for the last 5 months as part of my master's thesis.

I was trying to get the same results from one training round to the next one (same cfg, same dataset, etc, just running twice). However, I have failed so far :(

In order to do this I have added/modified the following lines in 'run_training' function in 'run_training.py':
```
random.seed(42)
torch.manual_seed(42)
np.random.seed(42)
if torch.cuda.is_available():
        torch.cuda.manual_seed_all(42)
        cudnn.deterministic = True
        cudnn.benchmark = False
```
This is done after the trainer instantiation, as it was in the original code (also tried before its instantiation with no difference).
In addition to this I tried to control the source of the variability during the batch generation, for this I modified the _LimitedLenWrapper_ class to use _MultiThreadedAugmenter_ instead of _NonDetMultiThreadedAugmenter_, and passed seeds to it when called in nnUNetTrainer.

However this two changes didn't work. First, I was checking the dice_per_class_per_region over the validation set, but also the training loss and the final dice I get after training over the complete validation set I have. In all those metrics there are differences across the different runs.

Any idea on how to solve this? I understand that running this in deterministic way can give me a misleading idea (with no noise) of my model's performance, but knowing that I want to control the behaviour so I can repeat my experiments and report the same numbers safely.

Thanks in advance,
Joaquín

**Comment by FabianIsensee** [comment]
Hey, good question. I have not attempted to make nnU-Net deterministic in a long time. From my understanding you already did everything you needed to do, so I am a bit at a loss myself. At this point I can only add this to the nnunet wish list and take a closer look once I find the time :-/ No promises though as to when this will be

**Comment by hdnminh** [comment]
Hi @FabianIsensee,

Why do you recommend us not using **the deterministic parameter**? I think it's really necessary to use it.
I mean if we don't use a deterministic parameter in training, maybe we could not reproduce the results that we reported.

Could you give me some thoughts on this?

Thank you so much.

**Comment by hdnminh** [comment]
Hi @FabianIsensee,

Could you help me please? Thank you so much.
